### ISSUE TITLE
Windows build issue

### ISSUE THREAD

**Issue by Thirumalai-Shaktivel** [lcompilers/lpython#284]
First error:
```console
$ cmake -DCMAKE_BUILD_TYPE=Debug -DWITH_LLVM=yes -DWITH_STACKTRACE=yes -DWITH_LFORTRAN_BINARY_MODFILES=no .
-- Building for: NMake Makefiles
CMake Error at CMakeLists.txt:12 (project):
  Running

   'nmake' '-?'

  failed with:

   The system cannot find the file specified


CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage
CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
-- Configuring incomplete, errors occurred!
See also "C:/Users/shock/OneDrive/Desktop/Open_Source/lpython/CMakeFiles/CMakeOutput.log".
```
which I fixed by appending `-G "MinGW Makefiles"` to the able command
Now I'm getting the following error:
```console
$ cmake -DCMAKE_BUILD_TYPE=Debug -DWITH_LLVM=yes -DWITH_STACKTRACE=yes -DWITH_LFORTRAN_BINARY_MODFILES=no . -G "MinGW Makefiles"
-- The C compiler identification is GNU 11.1.0
-- The CXX compiler identification is GNU 11.1.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/MinGW/bin/gcc.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/MinGW/bin/g++.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found ZLIB: C:/Users/shock/miniconda3/envs/lp/Library/lib/z.lib (found version "1.2.11")
-- Found LLVM 11.1.0
-- Using LLVMConfig.cmake in: C:/Users/shock/miniconda3/envs/lp/Library/lib/cmake/llvm
-- Found BFD: C:/MinGW/include
CMake Error at C:/Users/shock/miniconda3/envs/lp/Library/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find LINKH (missing: LINK_INCLUDE_DIR)
Call Stack (most recent call first):
  C:/Users/shock/miniconda3/envs/lp/Library/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  cmake/FindLINKH.cmake:4 (find_package_handle_standard_args)
  CMakeLists.txt:224 (find_package)


-- Configuring incomplete, errors occurred!
See also "C:/Users/shock/OneDrive/Desktop/Open_Source/lpython/CMakeFiles/CMakeOutput.log".
```
how do I fix this??

**Comment by Thirumalai-Shaktivel** [comment]
The Error is fixed I believe:
```console
$ cmake -DCMAKE_BUILD_TYPE=Debug -DWITH_LLVM=yes -DWITH_STACKTRACE=no -DWITH_LFORTRAN_BINARY_MODFILES=no . -G "MinGW Makefiles"
-- The C compiler identification is GNU 11.1.0
-- The CXX compiler identification is GNU 11.1.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/MinGW/bin/gcc.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/MinGW/bin/g++.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found ZLIB: C:/Users/shock/miniconda3/envs/lp/Library/lib/z.lib (found version "1.2.11")
-- Found LLVM 11.1.0
-- Using LLVMConfig.cmake in: C:/Users/shock/miniconda3/envs/lp/Library/lib/cmake/llvm


Configuration results
---------------------
LFORTRAN_VERSION: 0.1.0-922-g24d2f11de
CPACK_PACKAGE_FILE_NAME: lpython-0.1.0-922-g24d2f11de-Windows
C compiler      : C:/MinGW/bin/gcc.exe
C++ compiler    : C:/MinGW/bin/g++.exe
Build type: Debug
C compiler flags      : -g
C++ compiler flags    : -Wall -Wextra -g -ggdb
Installation prefix: C:/Program Files (x86)/lpython
WITH_LFORTRAN_ASSERT: yes
LFORTRAN_STATIC_BIN: no
WITH_STACKTRACE: no
WITH_UNWIND: no
WITH_BFD: no
WITH_DWARFDUMP: no
WITH_LINKH: no
WITH_MACHO: no
HAVE_LFORTRAN_DEMANGLE: yes
WITH_LLVM: yes
WITH_XEUS: no
WITH_JSON: no
WITH_FMT: no
WITH_LFORTRAN_BINARY_MODFILES: no
WITH_RUNTIME_LIBRARY: YES
WITH_TARGET_AARCH64: no
WITH_TARGET_X86: yes
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/shock/OneDrive/Desktop/Open_Source/lpython
```

**Comment by certik** [comment]
Yes, that looks good. See if it builds.

After that, can you help us fix `./run_tests.py` and also `integration_tests/run_tests.py` on Windows? The scripts currently do not run there. After it works, let's enable it at the CI.

**Comment by Thirumalai-Shaktivel** [comment]
Another issue:
```console
make -j8
[  1%] Building C object src/runtime/legacy/CMakeFiles/lfortran_runtime_static.dir/__/impure/lfortran_intrinsics.c.obj
[  3%] [  4%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/asr_to_py.cpp.objBuilding CXX object src/libasr/CMakeFiles/asr.dir/codegen/asr_to_cpp.cpp.obj
[  6%]
[  8%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/x86_assembler.cpp.obj[  9%]
Building CXX object src/libasr/CMakeFiles/asr.dir/pass/param_to_const.cpp.objBuilding CXX object src/libasr/CMakeFiles/asr.dir/codegen/asr_to_x86.cpp.obj

[ 13%] [ 13%] Building C object src/runtime/legacy/CMakeFiles/lfortran_runtime.dir/__/impure/lfortran_intrinsics.c.objBuilding CXX object src/tests/CMakeFiles/doctest.dir/doctest.cpp.obj

C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\runtime\impure\lfortran_intrinsics.c: In function '
s\shock\OneDrive\Desktop\Open_Source\lpython\src\runtime\impure\lfortran_intrinsics.c:
nction '':
_lfortran_i32sys_clockC:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\runtime\impure\lfortran_intrinsics.c:751:8:':
                                                                                                              implicit declaration of function '                                                        ' [
unction-declaration   ]
  751 |     if(
  751 |     if((CLOCK_MONOTONIC, &ts) == -1) {
      |
      |
^~~~~~~~~~~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\runtime\impure\lfortran_intrinsics.c:751:22:C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\runtime\impure\lfortran_intrinsics.c:751:22:                 '
TONIC
  751 |     if(clock_gettime(' undeclared (first use in this function)
  751 |     if(clock_gettime(CLOCK_MONOTONICCLOCK_MONOTONIC, &ts) == -1) {
      |                      , &ts) == -1) {
      |


\OneDrive\Desktop\Open_Source\lpython\src\runtime\impure\lfortran_intrinsics.c:751:22:
d identifier is reported only once for each function it appears in
each undeclared identifier is reported only once for each function it appears in

ve\Desktop\Open_Source\lpython\src\runtime\impure\lfortran_intrinsics.c:               In function '
ys_clock                      ':
':
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\runtime\impure\lfortran_intrinsics.c:771:22:C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\runtime\impure\lfortran_intrinsics.c:771:22:  error: error: ' CLOCK_MONOTONICCLOCK_MONOTONIC' undeclared (first use in this function)
  771 |     if(clock_gettime(
  771 |     if(clock_gettime(CLOCK_MONOTONICCLOCK_MONOTONIC, &ts) == -1) {
      |
      |                      ^~~~~~~~~~~~~~~^~~~~~~~~~~~~~~

make[2]: *** [src\runtime\legacy\CMakeFiles\lfortran_runtime_static.dir\build.make:75: src/runtime/legacy/CMakeFiles/lfortran_runtime_static.dir/__/impure/lfortran_intrinsics.c.obj] Error 1
make[2]: *** [src\runtime\legacy\CMakeFiles\lfortran_runtime.dir\build.make:75: src/runtime/legacy/CMakeFiles/lfortran_runtime.dir/__/impure/lfortran_intrinsics.c.obj] Error 1
make[1]: *** [CMakeFiles\Makefile2:476: src/runtime/legacy/CMakeFiles/lfortran_runtime_static.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles\Makefile2:450: src/runtime/legacy/CMakeFiles/lfortran_runtime.dir/all] Error 2
[ 14%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/for_all.cpp.obj
[ 16%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/do_loops.cpp.obj
[ 18%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/global_stmts.cpp.obj
In file included from C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.cpp:7:
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4646:14: error: 'mutex' in namespace 'std' does not name a type
 4646 |         std::mutex mutex;
      |              ^~~~~
In file included from C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.cpp:7:
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:2779:1: note: 'std::mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
 2778 | #include <io.h>
  +++ |+#include <mutex>
 2779 |
In file included from        C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.cpp:7Building CXX object src/libasr/CMakeFiles/asr.dir/pass/global_stmts_program.cpp.obj
:
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h: In member function 'virtual void doctest::{anonymous}::XmlReporter::test_case_exception(const doctest::TestCaseException&)':
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4798:34: error: 'mutex' is not a member of 'std'
 4798 |             std::lock_guard<std::mutex> lock(mutex);
      |                                  ^~~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4798:34: note: 'std::mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4798:39: error: template argument 1 is invalid
 4798 |             std::lock_guard<std::mutex> lock(mutex);
      |                                       ^
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4798:46: error: 'mutex' was not declared in this scope
 4798 |             std::lock_guard<std::mutex> lock(mutex);
      |                                              ^~~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4798:41: warning: unused variable 'lock'
-Wunused-variable]
 4798 |             std::lock_guard<std::mutex> lock(mutex);
      |                                         ^~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h: In member function 'virtual void doctest::{anonymous}::XmlReporter::subcase_start(const doctest::SubcaseSignature&)':
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4806:34: error: 'mutex' is not a member of 'std'
 4806 |             std::lock_guard<std::mutex> lock(mutex);
      |                                  ^~~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4806:34: note: 'std::mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4806:39: error: template argument 1 is invalid
 4806 |             std::lock_guard<std::mutex> lock(mutex);
      |                                       ^
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4806:46: error: 'mutex' was not declared in this scope
 4806 |             std::lock_guard<std::mutex> lock(mutex);
      |                                              ^~~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4806:41: warning: unused variable 'lock'
-Wunused-variable]
 4806 |             std::lock_guard<std::mutex> lock(mutex);
      |                                         ^~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h: In member function 'virtual void doctest::{anonymous}::XmlReporter::log_assert(const doctest::AssertData&)':
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4821:34: error: 'mutex' is not a member of 'std'
 4821 |             std::lock_guard<std::mutex> lock(mutex);
      |                                  ^~~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4821:34: note: 'std::mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4821:39: error: template argument 1 is invalid
 4821 |             std::lock_guard<std::mutex> lock(mutex);
      |                                       [ 21%] ^Building CXX object src/libasr/CMakeFiles/asr.dir/pass/select_case.cpp.obj

[ 22%] C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4821:46:Building CXX object src/libasr/CMakeFiles/asr.dir/pass/implied_do_loops.cpp.obj
 error: 'mutex' was not declared in this scope
 4821 |             std::lock_guard<std::mutex> lock(mutex);
      |                                              ^~~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4821:41: warning: unused variable 'lock' [-Wunused-variable]
 4821 |             std::lock_guard<std::mutex> lock(mutex);
      |                                         ^~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h: In member function 'virtual void doctest::{anonymous}::XmlReporter::log_message(const doctest::MessageData&)':
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4846:34: error: 'mutex' is not a member of 'std'
 4846 |             std::lock_guard<std::mutex> lock(mutex);
      |                                  ^~~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4846:34: note: 'std::mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4846:39: error: template argument 1 is invalid
 4846 |             std::lock_guard<std::mutex> lock(mutex);
      |                                       ^
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4846:46: error: 'mutex' was not declared in this scope
 4846 |             std::lock_guard<std::mutex> lock(mutex);
      |                                              ^~~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4846:41: warning: unused variable 'lock'
-Wunused-variable]
 4846 |             std::lock_guard<std::mutex> lock(mutex);
      |                                         ^~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h: At global scope:
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4889:14: error: 'mutex' in namespace 'std' does not name a type
 4889 |         std::mutex                    mutex;
      |              ^~~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:4889:9: note: 'std::mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
 4889 |         std::mutex                    mutex;
      |         ^~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h: In member function 'virtual void doctest::{anonymous}::ConsoleReporter::subcase_start(const doctest::SubcaseSignature&)':
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:5259:34: error: 'mutex' is not a member of 'std'
 5259 |             std::lock_guard<std::mutex> lock(mutex);
      |                                  ^~~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:5259:34: note: 'std::mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:5259:39: error: template argument 1 is invalid
 5259 |             std::lock_guard<std::mutex> lock(mutex);
      |                                       ^
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:5259:46: error: 'mutex' was not declared in this scope
 5259 |             std::lock_guard<std::mutex> lock(mutex);
      |                                              ^~~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:5259:41: warning: unused variable 'lock'
-Wunused-variable]
 5259 |             std::lock_guard<std::mutex> lock(mutex);
      |                                         ^~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h: In member function 'virtual void doctest::{anonymous}::ConsoleReporter::subcase_end()':
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:5265:34: error: 'mutex' is not a member of 'std'
 5265 |             std::lock_guard<std::mutex> lock(mutex);
      |                                  ^~~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:5265:34: note: 'std::mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:5265:39: error: template argument 1 is invalid
 5265 |             std::lock_guard<std::mutex> lock(mutex);
      |                                       ^
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:5265:46: error: 'mutex' was not declared in this scope
 5265 |             std::lock_guard<std::mutex> lock(mutex);
      |                                              ^~~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:5265:41: warning: unused variable 'lock'
-Wunused-variable]
 5265 |             std::lock_guard<std::mutex> lock(mutex);
      |                                         ^~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h: In member function 'virtual void doctest::{anonymous}::ConsoleReporter::log_assert(const doctest::AssertData&)':
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:5274:34: error: 'mutex' is not a member of 'std'
 5274 |             std::lock_guard<std::mutex> lock(mutex);
      |                                  ^~~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:5274:34: note: 'std::mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:5274:39: error: template argument 1 is invalid
 5274 |             std::lock_guard<std::mutex> lock(mutex);
      |                                       ^
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:5274:46: error: 'mutex' was not declared in this scope
 5274 |             std::lock_guard<std::mutex> lock(mutex);
      |                                              ^~~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:5274:41: warning: unused variable 'lock'
-Wunused-variable]
 5274 |             std::lock_guard<std::mutex> lock(mutex);
      |                                         ^~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h: In member function 'virtual void doctest::{anonymous}::ConsoleReporter::log_message(const doctest::MessageData&)':
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:5319:34: error: 'mutex' is not a member of 'std'
 5319 |             std::lock_guard<std::mutex> lock(mutex);
      |                                  ^~~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:5319:34: note: 'std::mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:5319:39: error: template argument 1 is invalid
 5319 |             std::lock_guard<std::mutex> lock(mutex);
      |                                       ^
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:5319:46: error: 'mutex' was not declared in this scope
 5319 |             std::lock_guard<std::mutex> lock(mutex);
      |                                              ^~~~~
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\tests\doctest.h:5319:41: warning: unused variable 'lock'
-Wunused-variable]
 5319 |             std::lock_guard<std::mutex> lock(mutex);
      |                                         ^~~~
[ 24%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/array_op.cpp.obj
[ 26%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/class_constructor.cpp.obj
[ 27%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/arr_slice.cpp.obj
make[2]: *** [src\tests\CMakeFiles\doctest.dir\build.make:75: src/tests/CMakeFiles/doctest.dir/doctest.cpp.obj] Error 1
make[1]: *** [CMakeFiles\Makefile2:312: src/tests/CMakeFiles/doctest.dir/all] Error 2
[ 29%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/print_arr.cpp.obj
[ 31%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/pass_utils.cpp.obj
[ 32%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/unused_functions.cpp.obj
[ 34%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/flip_sign.cpp.obj
[ 36%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/div_to_mul.cpp.obj
[ 37%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/fma.cpp.obj
[ 39%] [ 40%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/sign_from_value.cpp.objBuilding CXX object src/libasr/CMakeFiles/asr.dir/pass/inline_function_calls.cpp.obj

[ 42%] Building CXX object src/libasr/CMakeFiles/asr.dir/asr_verify.cpp.obj
[ 44%] Building CXX object src/libasr/CMakeFiles/asr.dir/asr_utils.cpp.obj
[ 45%] Building CXX object src/libasr/CMakeFiles/asr.dir/diagnostics.cpp.obj
[ 47%] Building CXX object src/libasr/CMakeFiles/asr.dir/stacktrace.cpp.obj
[ 49%] Building CXX object src/libasr/CMakeFiles/asr.dir/string_utils.cpp.obj
[ 50%] Building CXX object src/libasr/CMakeFiles/asr.dir/asr_scopes.cpp.obj
[ 52%] Building CXX object src/libasr/CMakeFiles/asr.dir/modfile.cpp.obj
[ 54%] Building CXX object src/libasr/CMakeFiles/asr.dir/serialization.cpp.obj
[ 55%] Building CXX object src/libasr/CMakeFiles/asr.dir/utils2.cpp.obj
[ 57%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/evaluator.cpp.obj
[ 59%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/asr_to_llvm.cpp.obj
[ 60%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/llvm_array_utils.cpp.obj
[ 62%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/llvm_utils.cpp.obj
[ 63%] Building CXX object src/libasr/CMakeFiles/asr.dir/pass/nested_vars.cpp.obj
In file included from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/ExecutionEngine.h:31,
                 from C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\libasr\codegen\asr_to_llvm.cpp:10:
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/Mutex.h:29:12: error: 'recursive_mutex' in namespace 'std' does not name a type
   29 |       std::recursive_mutex impl;
      |            ^~~~~~~~~~~~~~~
In file included from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/Mutex.h:19:1:C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/PassRegistry.h:23 ,
                 from note: C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/PassSupport.h:29',
                 from std::recursive_mutexC:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Pass.h:318
fined in header ',
                 from
forget to ',
                 from
ormInfo.h:25
   18 | #include <mutex>
  +++ |+,
                 from
ator.cpp:7
   19 |
:
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h:98:8: error: 'shared_mutex' in namespace 'std' does not name a type
   98 |   std::shared_mutex impl;
      |        ^~~~~~~~~~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h:21:1:C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/Mutex.h:  In member function 'note: bool llvm::sys::SmartMutex<mt_only>::lock()'':
std::shared_mutexC:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/Mutex.h:35:11:' is defined in header ' <shared_mutex>error: '; did you forget to ''#include <shared_mutex>impl'?
   20 | #include <shared_mutex>
  +++ |+' was not declared in this scope
   35 |           #include <shared_mutex>impl
   21 |
.lock();
      |           ^~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h: In member function 'bool llvm::sys::SmartRWMutex<mt_only>::lock_shared()
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/Mutex.h:C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h:112:7: In member function ' bool llvm::sys::SmartMutex<mt_only>::unlock()error: ':
'C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/Mutex.h:48:11:impl ' was not declared in this scope
  112 |       error: impl'.lock_shared();
      |       impl^~~~
   48 |
impl.unlock();
      |           ^~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/Mutex.h: In member function 'bool llvm::sys::SmartMutex<mt_only>::try_lock()C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h:':
 In member function 'C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/Mutex.h:62:18:bool llvm::sys::SmartRWMutex<mt_only>::unlock_shared() ':
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h:124:7:error:  'error: impl'' was not declared in this scope
   62 |           return implimpl.try_lock();
      |                  ' was not declared in this scope
  124 |       ^~~~impl
.unlock_shared();
      |       ^~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h: In member function 'bool llvm::sys::SmartRWMutex<mt_only>::lock()':
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h:137:7: error: 'impl' was not declared in this scope
  137 |       impl.lock();
      |       ^~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h: In member function 'bool llvm::sys::SmartRWMutex<mt_only>::unlock()':
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h:150:7: error: 'impl' was not declared in this scope
  150 |       impl.unlock();
      |       ^~~~
In file included from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/PassRegistry.h:23,
                 from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/PassSupport.h:29,
                 from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Pass.h:318,
                 from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Target/TargetMachine.h:19,
                 from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/ExecutionEngine.h:32,
                 from C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\libasr\codegen\asr_to_llvm.cpp:10:
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h: At global scope:
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h:98:8: error: 'shared_mutex' in namespace 'std' does not name a type
   98 |   std::shared_mutex impl;
      |        ^~~~~~~~~~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h:21:1: note: 'std::shared_mutex' is defined in header '<shared_mutex>'; did you forget to '#include <shared_mutex>'?
   20 | #include <shared_mutex>
  +++ |+#include <shared_mutex>
   21 |
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h: In member function 'bool llvm::sys::SmartRWMutex<mt_only>::lock_shared()':
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h:112:7: error: 'impl' was not declared in this scope
  112 |       impl.lock_shared();
      |       ^~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h: In member function 'bool llvm::sys::SmartRWMutex<mt_only>::unlock_shared()':
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h:124:7: error: 'impl' was not declared in this scope
  124 |       impl.unlock_shared();
      |       ^~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h: In member function 'bool llvm::sys::SmartRWMutex<mt_only>::lock()':
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h:137:7: error: 'impl' was not declared in this scope
  137 |       impl.lock();
      |       ^~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h: In member function 'bool llvm::sys::SmartRWMutex<mt_only>::unlock()':
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h:150:7: error: 'impl' was not declared in this scope
  150 |       impl.unlock();
      |       ^~~~
In file included from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/ExecutionEngine.h:31,
                 from C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\libasr\codegen\evaluator.cpp:9:
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/Mutex.h: At global scope:
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/Mutex.h:29:12: error: 'recursive_mutex' in namespace 'std' does not name a type
   29 |       std::recursive_mutex impl;
      |            ^~~~~~~~~~~~~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/Mutex.h:18:1: note: 'std::recursive_mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
   17 | #include <cassert>
  +++ |+#include <mutex>
   18 | #include <mutex>
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/Mutex.h: In member function 'bool llvm::sys::SmartMutex<mt_only>::lock()':
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/Mutex.h:35:11: error: 'impl' was not declared in this scope
   35 |           impl.lock();
      |           ^~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/Mutex.h: In member function 'bool llvm::sys::SmartMutex<mt_only>::unlock()':
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/Mutex.h:48:11: error: 'impl' was not declared in this scope
   48 |           impl.unlock();
      |           ^~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/Mutex.h: In member function 'bool llvm::sys::SmartMutex<mt_only>::try_lock()':
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/Mutex.h:62:18: error: 'impl' was not declared in this scope
   62 |           return impl.try_lock();
      |                  ^~~~
In file included from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/Core.h:20,
                 from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/Layer.h:16,
                 from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/IRCompileLayer.h:18,
                 from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/CompileUtils.h:1
,
                 from c:\users\shock\onedrive\desktop\open_source\lpython\src\libasr\codegen\kaleidoscopejit.h:20,
                 from C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\libasr\codegen\evaluator.cpp:53:
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/SymbolStringPool.h: At global scope:
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/SymbolStringPool.h:45:16: error: 'mute
' in namespace 'std' does not name a type
   45 |   mutable std::mutex PoolMutex;
      |                ^~~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/SymbolStringPool.h:1:1: note: 'std::mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
  +++ |+#include <mutex>
    1 | //===- SymbolStringPool.h - Multi-threaded pool for JIT symbols -*- C++ -*-===//
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/SymbolStringPool.h: In member function 'llvm::orc::SymbolStringPtr llvm::orc::SymbolStringPool::intern(llvm::StringRef)':
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/SymbolStringPool.h:156:24: error: 'mutex' is not a member of 'std'
  156 |   std::lock_guard<std::mutex> Lock(PoolMutex);
      |                        ^~~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/SymbolStringPool.h:156:24: note: 'std::mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/SymbolStringPool.h:156:29: error: template argument 1 is invalid
  156 |   std::lock_guard<std::mutex> Lock(PoolMutex);
      |                             ^
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/SymbolStringPool.h:156:36: error: 'PoolMutex' was not declared in this scope
  156 |   std::lock_guard<std::mutex> Lock(PoolMutex);
      |                                    ^~~~~~~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/SymbolStringPool.h: In member function 'void llvm::orc::SymbolStringPool::clearDeadEntries()':
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/SymbolStringPool.h:164:24: error: 'mutex' is not a member of 'std'
  164 |   std::lock_guard<std::mutex> Lock(PoolMutex);
      |                        ^~~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/SymbolStringPool.h:164:24: note: 'std::mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/SymbolStringPool.h:164:29: error: template argument 1 is invalid
  164 |   std::lock_guard<std::mutex> Lock(PoolMutex);
      |                             ^
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/SymbolStringPool.h:164:36: error: 'PoolMutex' was not declared in this scope
  164 |   std::lock_guard<std::mutex> Lock(PoolMutex);
      |                                    ^~~~~~~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/SymbolStringPool.h: In member function 'bool llvm::orc::SymbolStringPool::empty() const':
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/SymbolStringPool.h:173:24: error: 'mutex' is not a member of 'std'
  173 |   std::lock_guard<std::mutex> Lock(PoolMutex);
      |                        ^~~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/SymbolStringPool.h:173:24: note: 'std::mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/SymbolStringPool.h:173:29: error: template argument 1 is invalid
  173 |   std::lock_guard<std::mutex> Lock(PoolMutex);
      |                             ^
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/SymbolStringPool.h:173:36: error: 'PoolMutex' was not declared in this scope
  173 |   std::lock_guard<std::mutex> Lock(PoolMutex);
      |                                    ^~~~~~~~~
In file included from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/Layer.h:16,
                 from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/IRCompileLayer.h:18,
                 from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/CompileUtils.h:1
,
                 from c:\users\shock\onedrive\desktop\open_source\lpython\src\libasr\codegen\kaleidoscopejit.h:20,
                 from C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\libasr\codegen\evaluator.cpp:53:
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/Core.h: At global scope:
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/Core.h:1278:16: error: 'recursive_mute
' in namespace 'std' does not name a type
 1278 |   mutable std::recursive_mutex SessionMutex;
      |                ^~~~~~~~~~~~~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/Core.h:21:1: note: 'std::recursive_mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
   20 | #include "llvm/ExecutionEngine/Orc/SymbolStringPool.h"
  +++ |+#include <mutex>
   21 | #include "llvm/ExecutionEngine/OrcV1Deprecation.h"
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/Core.h:1290:16: error: 'recursive_mute
' in namespace 'std' does not name a type
 1290 |   mutable std::recursive_mutex OutstandingMUsMutex;
      |                ^~~~~~~~~~~~~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/Core.h:1290:11: note: 'std::recursive_mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
 1290 |   mutable std::recursive_mutex OutstandingMUsMutex;
      |           ^~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/Core.h: In member function 'decltype(auto) llvm::orc::ExecutionSession::runSessionLocked(Func&&)':
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/Core.h:1121:26: error: 'recursive_mute
' is not a member of 'std'
 1121 |     std::lock_guard<std::recursive_mutex> Lock(SessionMutex);
      |                          ^~~~~~~~~~~~~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/Core.h:1121:26: note: 'std::recursive_mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/Core.h:1121:41: error: template argument 1 is invalid
 1121 |     std::lock_guard<std::recursive_mutex> Lock(SessionMutex);
      |                                         ^
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/Core.h:1121:48: error: 'SessionMutex' was not declared in this scope
 1121 |     std::lock_guard<std::recursive_mutex> Lock(SessionMutex);
      |                                                ^~~~~~~~~~~~
In file included from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/Mangling.h:17,
                 from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/Layer.h:17,
                 from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/IRCompileLayer.h:18,
                 from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/CompileUtils.h:1
,
                 from c:\users\shock\onedrive\desktop\open_source\lpython\src\libasr\codegen\kaleidoscopejit.h:20,
                 from C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\libasr\codegen\evaluator.cpp:53:
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/ThreadSafeModule.h: At global scope:
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/ThreadSafeModule.h:35:10: error: 'recursive_mutex' in namespace 'std' does not name a type
   35 |     std::recursive_mutex Mutex;
      |          ^~~~~~~~~~~~~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/ThreadSafeModule.h:1:1: note: 'std::recursive_mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
  +++ |+#include <mutex>
    1 | //===----------- ThreadSafeModule.h -- Layer interfaces ---------*- C++ -*-===//
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/ThreadSafeModule.h:46:27: error: 'recursive_mutex' is not a member of 'std'
   46 |     std::unique_lock<std::recursive_mutex> L;
      |                           ^~~~~~~~~~~~~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/ThreadSafeModule.h:46:27: note: 'std::recursive_mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/ThreadSafeModule.h:46:42: error: template argument 1 is invalid
   46 |     std::unique_lock<std::recursive_mutex> L;
      |                                          ^
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/ThreadSafeModule.h: In constructor 'llvm::orc::ThreadSafeContext::Lock::Lock(std::shared_ptr<llvm::orc::ThreadSafeContext::State>)':
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/ThreadSafeModule.h:42:66: error: 'using element_type = struct llvm::orc::ThreadSafeContext::State' {aka 'struct llvm::orc::ThreadSafeContext::State'} has no member named 'Mutex'
   42 |     Lock(std::shared_ptr<State> S) : S(std::move(S)), L(this->S->Mutex) {}
      |                                                                  ^~~~~
In file included from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/CompileUtils.h:1
,
                 from c:\users\shock\onedrive\desktop\open_source\lpython\src\libasr\codegen\kaleidoscopejit.h:20,
                 from C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\libasr\codegen\evaluator.cpp:53:
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/IRCompileLayer.h: At global scope:
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/IRCompileLayer.h:61:16: error: 'mutex' in namespace 'std' does not name a type
   61 |   mutable std::mutex IRLayerMutex;
      |                ^~~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/IRCompileLayer.h:19:1: note: 'std::mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
   18 | #include "llvm/ExecutionEngine/Orc/Layer.h"
  +++ |+#include <mutex>
   19 | #include "llvm/Support/Error.h"
In file included from c:\users\shock\onedrive\desktop\open_source\lpython\src\libasr\codegen\kaleidoscopejit.h:23,
                 from C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\libasr\codegen\evaluator.cpp:53:
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/RTDyldObjectLinkingLayer.h:137:16: error: 'mutex' in namespace 'std' does not name a type
  137 |   mutable std::mutex RTDyldLayerMutex;
      |                ^~~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/Orc/RTDyldObjectLinkingLayer.h:29:1: note:
'std::mutex' is defined in header '<mutex>'; did you forget to '#include <mutex>'?
   28 | #include <cassert>
  +++ |+#include <mutex>
   29 | #include <functional>
In file included from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/PassRegistry.h:23,
                 from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/PassSupport.h:29,
                 from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Pass.h:318,
                 from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/IR/PassManager.h:48,
                 from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/IR/Verifier.h:24,
                 from C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\libasr\pass\nested_vars.cpp:7:
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h:98:8: error: 'shared_mutex' in namespace 'std' does not name a type
   98 |   std::shared_mutex impl;
      |        ^~~~~~~~~~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h:21:1: note: 'std::shared_mutex' is defined in header '<shared_mutex>'; did you forget to '#include <shared_mutex>'?
   20 | #include <shared_mutex>
  +++ |+#include <shared_mutex>
   21 |
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h: In member function 'bool llvm::sys::SmartRWMutex<mt_only>::lock_shared()':
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h:112:7: error: 'impl' was not declared in this scope
  112 |       impl.lock_shared();
      |       ^~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h: In member function 'bool llvm::sys::SmartRWMutex<mt_only>::unlock_shared()':
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h:124:7: error: 'impl' was not declared in this scope
  124 |       impl.unlock_shared();
      |       ^~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h: In member function 'bool llvm::sys::SmartRWMutex<mt_only>::lock()':
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h:137:7: error: 'impl' was not declared in this scope
  137 |       impl.lock();
      |       ^~~~
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h: In member function 'bool llvm::sys::SmartRWMutex<mt_only>::unlock()':
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Support/RWMutex.h:150:7: error: 'impl' was not declared in this scope
  150 |       impl.unlock();
      |       ^~~~
make[2]: *** [src\libasr\CMakeFiles\asr.dir\build.make:601: src/libasr/CMakeFiles/asr.dir/pass/nested_vars.cpp.obj] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [src\libasr\CMakeFiles\asr.dir\build.make:556: src/libasr/CMakeFiles/asr.dir/codegen/asr_to_llvm.cpp.obj] Error 1
make[2]: *** [src\libasr\CMakeFiles\asr.dir\build.make:541: src/libasr/CMakeFiles/asr.dir/codegen/evaluator.cpp.obj] Error 1
make[1]: *** [CMakeFiles\Makefile2:286: src/libasr/CMakeFiles/asr.dir/all] Error 2
make: *** [Makefile:165: all] Error 2
```

**Comment by certik** [comment]
Yes, you'll have to fix the header files or remove the code that doesn't run on Windows.

**Comment by Thirumalai-Shaktivel** [comment]
This[ issue](https://github.com/lcompilers/lpython/issues/284#issuecomment-1079221099) is fixed by installing [MinGW-w64](https://www.mingw-w64.org/) (https://github.com/lcompilers/lpython/issues/284#issuecomment-1079221099), and making it as a main C++ compiler

**Comment by Thirumalai-Shaktivel** [comment]
The let-out issues are:
```console
C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\libasr\utils2.cpp:75:13: error: 'ENABLE_VIRTUAL_TERMINAL_PROCESSING' was not declared in this scope
     mode |= ENABLE_VIRTUAL_TERMINAL_PROCESSING;
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [src\libasr\CMakeFiles\asr.dir\build.make:526: src/libasr/CMakeFiles/asr.dir/utils2.cpp.obj] Error 1
make[2]: *** Waiting for unfinished jobs....
```
```console
In file included from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Object/ObjectFile.h:23,
                 from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/DebugInfo/DIContext.h:18,
                 from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/RuntimeDyld.h:19,
                 from C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/ExecutionEngine/JITEventListener.h:19,
                 from C:/Users/shock/OneDrive/Desktop/Open_Source/lpython/src/libasr/codegen/KaleidoscopeJIT.h:23,
                 from C:\Users\shock\OneDrive\Desktop\Open_Source\lpython\src\libasr\codegen\evaluator.cpp:53:
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Object/SymbolicFile.h: In function 'OStream& llvm::object::operator<<(OStream&, const llvm::object::DataRefImpl&)':
C:/Users/shock/miniconda3/envs/lp/Library/include/llvm/Object/SymbolicFile.h:47:30: error: expected ')' before 'PRIxPTR'
   OS << "(" << format("0x%08" PRIxPTR, D.p) << " (" << format("0x%08x", D.d.a)
                      ~       ^~~~~~~~
                              )
make[2]: *** [src\libasr\CMakeFiles\asr.dir\build.make:541: src/libasr/CMakeFiles/asr.dir/codegen/evaluator.cpp.obj] Error 1
make[1]: *** [CMakeFiles\Makefile2:286: src/libasr/CMakeFiles/asr.dir/all] Error 2
make: *** [Makefile:165: all] Error 2
```

**Comment by certik** [comment]
The first one depends on Windows headers. Simply define in that file to use "linux".

The second issue I've seen many times --- it's some  kind of a define in a header that messes up the code: https://github.com/tensorflow/tensorflow/issues/12998, and the solution I think is to define `#define __STDC_FORMAT_MACROS ` before importing that LLVM header. Possibly in our CMake.

**Comment by Thirumalai-Shaktivel** [comment]
I included `#define __STDC_FORMAT_MACROS` in `asr_utils.h`, the second issue was fixed!

**Comment by Thirumalai-Shaktivel** [comment]
This is the correct fix: I added `add_definitions("-D__STDC_FORMAT_MACROS")` in CMAKELists.txt in `libasr/`, which fixed the issue

**Comment by Thirumalai-Shaktivel** [comment]
`Simply define in that file to use "linux"` I didn't get this??

**Comment by certik** [comment]
The correct fix will be to only do it when it is needed, not always, but just get everything working for you on Windows and then submit a PR and we'll polish it.

Regarding utils2, in that function you can't use `_WIN32`, as it assumes MSVC, but you don't have MSVC and don't have the `ENABLE_VIRTUAL_TERMINAL_PROCESSING` thing defined on your system. You would if you used MSVC, but apparently not when using mingw.

**Comment by Thirumalai-Shaktivel** [comment]
I will make a PR we will polish there, Thank you!
